File: src/searching/cpp/jump_search/src/jump_search.cpp
#include "jump_search.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <optional>
#include <vector>

template <typename T>
std::optional<std::size_t> jump_search(const std::vector<T> &arr, const T &target) {
  std::size_t n = arr.size();
  if (n == 0) {
    return std::nullopt;
  }

  std::size_t step = static_cast<std::size_t>(std::sqrt(static_cast<double>(n)));
  std::size_t prev = 0;

  // Jump ahead until we find a block where target could be
  while (arr[std::min(step, n) - 1] < target) {
    prev = step;
    step += static_cast<std::size_t>(std::sqrt(static_cast<double>(n)));
    if (prev >= n) {
      return std::nullopt;
    }
  }

  // Linear search within the block
  while (arr[prev] < target) {
    ++prev;
    if (prev == std::min(step, n)) {
      return std::nullopt;
    }
  }

  if (arr[prev] == target) {
    return prev;
  }

  return std::nullopt;
}

// Explicit template instantiations
template std::optional<std::size_t> jump_search<int>(const std::vector<int> &arr, const int &target);
template std::optional<std::size_t> jump_search<float>(const std::vector<float> &arr, const float &target);
template std::optional<std::size_t> jump_search<double>(const std::vector<double> &arr, const double &target);
template std::optional<std::size_t> jump_search<char>(const std::vector<char> &arr, const char &target);

File: src/searching/python/jump_search/src/jump_search.py
"""Jump search algorithm implementation."""

import math
from typing import List, Optional, TypeVar

T = TypeVar("T")


def jump_search(array: List[T], target: T) -> Optional[int]:
    """
    Search for a target value in a sorted array using jump search.

    This algorithm jumps ahead by fixed block sizes to find the block that may
    contain the target, then performs a linear scan inside that block.

    Time complexity: O(√n)
    Space complexity: O(1)

    Args:
        array: The sorted list of comparable elements to search.
        target: The value to search for.

    Returns:
        The index of the target if found, None otherwise.

    Note:
        The array must be sorted in ascending order for correct results.
    """
    n = len(array)
    if n == 0:
        return None

    step = int(math.sqrt(n))
    prev = 0

    # Jump ahead until we find a block where target could be
    while array[min(step, n) - 1] < target:
        prev = step
        step += int(math.sqrt(n))
        if prev >= n:
            return None

    # Linear search within the block
    while array[prev] < target:
        prev += 1
        if prev == min(step, n):
            return None

    if array[prev] == target:
        return prev

    return None
